### ISSUE TITLE
events not called if WorldObject material color has an alpha value less than 1.0

### ISSUE THREAD

**Issue by kushalkolar** [pygfx/pygfx#474]
Simplest way to reproduce this is with the `multi_selection.py` example. Set the color for `default_material` on line 35 to something with an alpha < 1.0, and events are not called. It is true for lines as well, so I'm guessing this happens across many types of objects.

`wgpu-py` version is `0.9.4`, pygfx version is the current head of `main` on github.

```python
from functools import partial
from random import randint, random

from wgpu.gui.auto import WgpuCanvas, run
import pygfx as gfx


canvas = WgpuCanvas()
renderer = gfx.renderers.WgpuRenderer(canvas)

camera = gfx.PerspectiveCamera(70, 16 / 9)
camera.position.z = 400

controller = gfx.OrbitController(camera.position.clone())
controller.add_default_event_handlers(renderer, camera)

scene = gfx.Scene()
scene.add(gfx.AmbientLight(), camera.add(gfx.DirectionalLight()))

geometry = gfx.box_geometry(40, 40, 40)
default_material = gfx.MeshPhongMaterial(color=(1., 0., 1., 0.2))
selected_material = gfx.MeshPhongMaterial(color="#FF0000")
hover_material = gfx.MeshPhongMaterial(color="#FFAA00")

outline = gfx.BoxHelper(thickness=3, color="#fa0")
scene.add(outline)

selected_objects = []


def set_material(material, obj):
    if isinstance(obj, gfx.Mesh):
        obj.material = material


def select(event):
    # when this event handler is invoked on non-leaf nodes of the
    # scene graph, event.target will still point to the leaf node that
    # originally triggered the event, so we use event.current_target
    # to get a reference to the node that is currently handling
    # the event, which can be a Mesh, a Group or None (the event root)
    obj = event.current_target

    # prevent propagation so we handle this event only at one
    # level in the hierarchy
    event.stop_propagation()

    # clear selection
    if selected_objects and "Shift" not in event.modifiers:
        while selected_objects:
            ob = selected_objects.pop()
            ob.traverse(partial(set_material, default_material))

    # if the background was clicked, we're done
    if isinstance(obj, gfx.Renderer):
        return

    # set selection (group or mesh)
    selected_objects.append(obj)
    obj.traverse(partial(set_material, selected_material))


def hover(event):
    obj = event.target
    if event.type == "pointer_enter":
        obj.add(outline)
        outline.set_transform_by_object(obj, "local", scale=1.1)
    elif outline.parent:
        outline.parent.remove(outline)


def random_rotation():
    return gfx.linalg.Quaternion().set_from_euler(
        gfx.linalg.Euler(
            (random() - 0.5) / 100, (random() - 0.5) / 100, (random() - 0.5) / 100
        )
    )


def animate():
    def random_rot(obj):
        if hasattr(obj, "random_rotation"):
            obj.rotation.multiply(obj.random_rotation)

    scene.traverse(random_rot)
    controller.update_camera(camera)
    renderer.render(scene, camera)
    canvas.request_draw()


if __name__ == "__main__":
    renderer.add_event_handler(select, "click")

    # Build up scene
    for _ in range(4):
        group = gfx.Group()
        group.random_rotation = random_rotation()
        group.add_event_handler(select, "double_click")
        scene.add(group)

        for _ in range(10):
            cube = gfx.Mesh(geometry, default_material)
            cube.position.x = randint(-200, 200)
            cube.position.y = randint(-200, 200)
            cube.position.z = randint(-200, 200)
            cube.random_rotation = random_rotation()
            cube.add_event_handler(select, "click")
            cube.add_event_handler(hover, "pointer_enter", "pointer_leave")
            group.add(cube)

    canvas.request_draw(animate)
    run()
```

**Comment by Korijn** [comment]
This is a limitation of the current picking implementation. When objects are rendered, their ids are written to a second buffer on a per pixel basis. There's room for only one entry per pixel. The front-most opaque object ends up in there.

There are edge cases that this algorithm has by design which are hard to resolve, so as far as I can remember we decided to exclude transparent objects from the picking buffer. For example, some of these edge cases are as follows:

Let's say that you have multiple overlapping transparant objects occupying a pixel, which one should be stored in the buffer? The front-most surface, or the least transparent one, or should some other ranking metric be used?

Even worse, let's say there's two transparent planes intersecting eachother inside a specific pixel. Now there is no clear winner anymore when considering which one would be "front-most".

Long story short, it depends on the application what kind of picking algorithm is best. We went with the painters algorithm here since it runs "for free" on the GPU as a side effect of rendering, but you could also implement picking on the CPU using ray casting against the geometry in the scene. We have not implemented this though.

**Comment by almarklein** [comment]
Yes, we decided that only opaque objects are pickable. 

During the render pass that draws the transparent fragments, the picking texture is not attached. So it would not be easy to force certain objects to write their picking info regardless of their alpha.

@kushalkolar do you have particular use-cases where this is a problem, or can you (with this knowledge) work around it?

**Comment by kushalkolar** [comment]
Thanks for the details! I think we discussed it briefly last time but I understand it better now.

The use is moving a rectangular selector with the mouse. I could make a box using only opaque lines but it is easier to click and drag a rectangular box which has a fill color set to zero or low alpha in addition to lines on the edges. I guess the best way around it would be to add the mouse events to the renderer, and then check if any selector instances are under that click event?

**Comment by sharon92** [comment]
@almarklein 

> During the render pass that draws the transparent fragments, the picking texture is not attached. So it would not be easy to force certain objects to write their picking info regardless of their alpha.

I would like to try to pick transparent fragments. Do you have some tips how I could achieve that? 
In wgpu\_blender.py -> OpaquePass -> get_shader_code, if alpha is not 1 its discarded. I am guessing here some special things need to be done?

I am trying to pick faces of a wireframe mesh (can only pick edges), by overlaying with the same mesh and alpha=0

**Comment by almarklein** [comment]
You could try, but I fear that the result would be unsatisfactory. The idea with (semi) transparent fragments is that they do not write to the depth buffer, so any fragment can be drawn in any order. So if the shader would write to the picking texture for transparent fragments, it will do so in an undetermined order and there is no way to make it so that e.g. you'd pick the frontmost layer.

Writing this down, there might be one solution, and that'd be using the `weighted_plus` blend mode. That mode has one render pass dedicated to rendering the frontmost transparent pixels. I'll have a look at this option.

**Comment by sharon92** [comment]
@almarklein in my use case, I am not overlaying different meshes on each other, so the picked face would be unique, i.e. it doesn't matter if it's not the frontmost or backmost.

**Comment by almarklein** [comment]
https://github.com/pygfx/pygfx/pull/555
